[MXNET-979] Add fix_beta support in BatchNorm

sandeep-krishnamurthy opened the pull request:
## Description ##
Current implementation of Batchnorm does not support ignoring the beta parameter. However, it does support fixing the gamma parameter via 'fix_gamma' boolean parameter, default to True.

In this PR:
1. Add new option in Batchnorm 'fix_beta' and update batchnorm operator to support ignoring the beta in forward and backward calculation.
2. Gluon had a bug - Takes a boolean parameter 'center' and 'scale' where center=Beta and scale=Gamma. However, behind the scene it was ignoring 'center' parameter as symbolic API did not had that support. This PR fixes it to pass fix_beta param.
3. By default, fix_beta=False for backward compatibility.

Fixes #11774
This will also be useful in Keras-MXNet as Keras BatchNorm layer has Center=True/False param.

## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [X] Support fix_beta in Batchnorm
- [X] Support center=False in Gluon
- [ ] Tests

## Comments ##
Testing done:
1. Tests are passing.
2. Tested one end to end example - https://gluon.mxnet.io/chapter04_convolutional-neural-networks/cnn-batch-norm-gluon.html

@roywei, @eric-haibin-lin - <s>Can you please provide your initial comments? While, I work on GPU
 tests </s> 
This is ready for review.

stu1130 commented:
Thanks for your contribution! @sandeep-krishnamurthy 
@mxnet-label-bot[pr-work-in-progress]

sandeep-krishnamurthy commented:
@eric-haibin-lin @zhreshold - Can you please provide your feedback on this change? Thanks.

vandanavk commented:
Please add 

> Fixes https://github.com/apache/incubator-mxnet/issues/11774

 in the PR description. It will automatically close the issue when the PR is merged.

zhreshold commented:
LGTM. Since BN is super important, I would like to confirm it will pass some training test before get merged into master.

sandeep-krishnamurthy commented:
@szha - cudnn batchnorm is covered in this PR. 
Since SyncBatchNorm is independent of BatchNorm operator, I would prefer to do it separately. Created a feature request issue to follow up. https://github.com/apache/incubator-mxnet/issues/12728

sandeep-krishnamurthy commented on src/operator/nn/batch_norm.cc:
```diff
@@ -164,18 +164,42 @@ void BatchNormForwardImpl(mshadow::Stream<cpu> *,
+        // Case 3
+        // fix_gamma = True
+        // fix_beta = False
       if (IsBNWriting(req[batchnorm::kData])) {
```
fixed.

sandeep-krishnamurthy commented on tests/python/gpu/test_operator_gpu.py:
```diff
@@ -303,35 +303,35 @@ def test_batchnorm_with_type():
 
   # V2, 2D
-  sym = mx.sym.BatchNorm(name='norm', fix_gamma=False, cudnn_off=True)
+  sym = mx.sym.BatchNorm(name='norm', fix_gamma=False, fix_beta=False, cudnn_off=True)
```
Thanks. good catch. added tests for missing fix_beta and fix_gamma params.

zhreshold commented on src/operator/nn/batch_norm.cc:
```diff
@@ -164,25 +164,65 @@ void BatchNormForwardImpl(mshadow::Stream<cpu> *,
-                      *out_data = static_cast<DType>(
-                        ((*in_data - thisMean) * thisInvstd) * thisWeight + thisBias);
-                    });
+      if (!param_.fix_beta) {
```
Do you really need 4 cases rather than collate 2 cases separately?

sandeep-krishnamurthy replied:
It is basically 2 cases based on fix_gamma. fix_beta cases are collated within it.
This has more readability with flags on the cases. Suggestions?

zhreshold replied:
Fix gamma and beta only involves whether including (*gamma) and (+bias), but here's more than 50 lines only to accomplish that. It should be fine with the current lambda function with one or two cases, but with 4 cases or maybe 8 cases in the future, it's going to cause a lot of trouble.

sandeep-krishnamurthy commented on src/operator/nn/batch_norm.cc:
```diff
@@ -164,18 +164,42 @@ void BatchNormForwardImpl(mshadow::Stream<cpu> *,
+        // fix_beta = False
+        if (IsBNWriting(req[batchnorm::kData])) {
+          ForEachFast(inputData, outputData, channel,
+                      [thisWeight, thisBias, thisMean, thisInvstd](const DType *in_data,
```
As discussed offline, this change would make readability difficult and pass more params unnecessarily across functions.

sandeep-krishnamurthy commented:
@apeforest @szha @zhreshold - Thanks for the review. Addressed your comments.
@zhreshold - Tested with and without scale factor with end-to-end ResNet50 on CIFAR10 example. Works as expected.

sandeep-krishnamurthy commented:
@zhreshold - Simplified the logic to set defaults 0, 1 for gamma, beta respectively, when asked to ignore gamma / beta.
